v4l2-decoder-conformance: Enable Fluster tests on ChromeOS

denisyuji opened the pull request:
This PR combines creates a new template `v4l2-decoder-conformance-chromeos` to execute Fluster test on ChromeOS.
It re-use the ChromeOS kernel module loading structure from Tast and Fluster parser script calling from v4l2-decoder-conformance test.
This PR depends on the template defined on https://github.com/kernelci/kernelci-core/pull/2617 PR

laura-nao commented:
Added `staging-skip` label while the rootfs image URL gets fixed.

laura-nao commented:
@denisyuji did you try running a manual LAVA job with these changes? might be useful as a reference to check the templates :)

denisyuji commented:
> @denisyuji did you try running a manual LAVA job with these changes? might be useful as a reference to check the templates :)

Hi @laura-nao . Yes, I did! I replaced the variables by the included templates and got this lava job definition. I was able to run it correctly

```
job_name: v4l2-decoder-conformance-chromeos-vp9-VP9-TEST-VECTORS-GStreamer-VP9-V4L2SL-Gst1.0-mt8195-cherry-tomato-r2

device_type: mt8195-cherry-tomato-r2

visibility: public

priority: high

context:
  test_character_delay: 10

timeouts:
  action:
    minutes: 10
  actions:
    power-off:
      seconds: 30
  job:
    minutes: 30
  queue:
    days: 2

metadata:
  node_id: 9f30111364a86e3d05592e3b
  api_config_name: staging
  storage_config_name: staging-azure

actions:
- deploy:
    dtb:
      url: https://storage.chromeos.kernelci.org/images/kernel/v6.1-mediatek/dtbs/mediatek/mt8195-cherry-tomato-r2.dtb
    kernel:
      url: https://storage.chromeos.kernelci.org/images/kernel/v6.1-mediatek/kernel/Image
    modules:
      compression: xz
      url: https://storage.chromeos.kernelci.org/images/kernel/v6.1-mediatek/modules.tar.xz
    namespace: modules
    nfsrootfs:
      compression: xz
      url: https://storage.chromeos.kernelci.org/images/rootfs/debian/bookworm-cros-flash/20240422.0/arm64/full.rootfs.tar.xz
    os: oe
    ramdisk:
      compression: gz
      url: https://storage.chromeos.kernelci.org/images/rootfs/debian/bookworm-cros-flash/20240422.0/arm64/initrd.cpio.gz
    timeout:
      minutes: 10
    to: tftp
- boot:
    commands: nfs
    namespace: modules
    method: depthcharge
    failure_retry: 3
    prompts:
    - '/ #'
    timeout:
      minutes: 10
    timeouts:
      bootloader-commands:
        minutes: 10
      auto-login-action:
        minutes: 10

- test:
    namespace: modules
    timeout:
      minutes: 5
    definitions:
    - repository:
        metadata:
          format: Lava-Test Test Definition 1.0
          name: modules
          description: modules
          os:
            - oe
          scope:
            - functional
          environment:
            - lava-test-shell
        run:
          steps:
          - /opt/chromeos/install-modules 'https://kciapistagingstorage1.file.core.windows.net/early-access/kbuild-gcc-12-arm64-chromeos-mediatek-66995b6baa07c494f8d68b19/modules.tar.xz?sv=2022-11-02&ss=f&srt=sco&sp=r&se=2024-10-17T19:19:12Z&st=2023-10-17T11:19:12Z&spr=https&sig=sLmFlvZHXRrZsSGubsDUIvTiv%2BtzgDq6vALfkrtWnv8%3D'
            || lava-test-raise "modules-install"
      from: inline
      name: modules
      path: inline/modules.yaml

- deploy:
    dtb:
      url: https://kciapistagingstorage1.file.core.windows.net/early-access/kbuild-gcc-12-arm64-chromeos-mediatek-66995b6baa07c494f8d68b19/dtbs/mediatek/mt8195-cherry-tomato-r2.dtb?sv=2022-11-02&ss=f&srt=sco&sp=r&se=2024-10-17T19:19:12Z&st=2023-10-17T11:19:12Z&spr=https&sig=sLmFlvZHXRrZsSGubsDUIvTiv%2BtzgDq6vALfkrtWnv8%3D
    kernel:
      url: https://kciapistagingstorage1.file.core.windows.net/early-access/kbuild-gcc-12-arm64-chromeos-mediatek-66995b6baa07c494f8d68b19/Image?sv=2022-11-02&ss=f&srt=sco&sp=r&se=2024-10-17T19:19:12Z&st=2023-10-17T11:19:12Z&spr=https&sig=sLmFlvZHXRrZsSGubsDUIvTiv%2BtzgDq6vALfkrtWnv8%3D
    namespace: chromeos
    os: oe
    timeout:
      minutes: 10
    to: tftp

- boot:
    namespace: chromeos
    timeout:
      minutes: 10
    timeouts:
      bootloader-commands:
        minutes: 10
      auto-login-action:
        minutes: 10
    failure_retry: 3
    method: depthcharge
    commands: emmc
    extra_kernel_args: cros_secure cros_debug root=PARTUUID=566f7961-6765-7220-746f-20756e697665 rootwait ro
    prompts:
      - 'localhost(.*)~(.*)#'
    failure_message: 'chromeos-boot-alert: self_repair'
    auto_login:
      login_prompt: "login:"
      username: "root"
      password_prompt: "Password:"
      password: "test0000"

- test:
    namespace: chromeos
    timeout:
      minutes: 30
    docker:
      image: kernelci/cros-tast
      wait:
        device: true
    results:
      location: /home/cros/lava
    definitions:
    - from: inline
      name: v4l2-test
      path: inline/v4l2-test.yaml
      repository:
        metadata:
          format: Lava-Test Test Definition 1.0
          name: cros-tast
        run:
          steps:
            - cd /home/cros
            - lava-test-set start setup
            - lava-test-case tast-tarball --shell curl -s 'https://storage.chromeos.kernelci.org/images/rootfs/chromeos/chromiumos-cherry/20240514.0/arm64/tast.tgz'
              \| tar xzvf - && cp remote_test_runner /usr/bin/remote_test_runner &&
              mkdir -p /usr/libexec/tast/bundles/remote/ && cp cros /usr/libexec/tast/bundles/remote/
            - for i in $(seq 1 60); do ping -c 1 -w 1 $(lava-target-ip) && break || sleep 1; done
            - ping -c 1 -w 1 $(lava-target-ip) || lava-test-raise "cros-device-unreachable"
            - lava-test-case touch1-result --shell
              ./ssh_retry.sh
              -o StrictHostKeyChecking=no
              -o UserKnownHostsFile=/dev/null
              -i /home/cros/.ssh/id_rsa
              root@$(lava-target-ip)
              touch /home/chronos/results.xml
            - lava-test-case touch2-result --shell
              ./ssh_retry.sh
              -o StrictHostKeyChecking=no
              -o UserKnownHostsFile=/dev/null
              -i /home/cros/.ssh/id_rsa
              root@$(lava-target-ip)
              touch /home/root/results.xml
            - lava-test-case touch3-result --shell
              ./ssh_retry.sh
              -o StrictHostKeyChecking=no
              -o UserKnownHostsFile=/dev/null
              -i /home/cros/.ssh/id_rsa
              root@$(lava-target-ip)
              touch /home/user/results.xml
            - ./ssh_retry.sh
              -o StrictHostKeyChecking=no
              -o UserKnownHostsFile=/dev/null
              -i /home/cros/.ssh/id_rsa
              root@$(lava-target-ip)
              python3 /usr/bin/fluster_parser.py -ts VP9-TEST-VECTORS -d GStreamer-VP9-V4L2SL-Gst1.0
```

laura-nao commented:
> > @denisyuji did you try running a manual LAVA job with these changes? might be useful as a reference to check the templates :)
> 
> Hi @laura-nao . Yes, I did! I replaced the variables on the included templates and got the following lava job definition. I was able to run it correctly

Great, thanks! I believe we can slim down the test action and remove the unnecessary parts. We can probably start by running the test directly on the DUT (no real need to run it from the docker container, which is mainly used for Tast tests) and by skipping the tast tarball downloading/unpacking. Could you please try running the manual LAVA job again with a slimmed down version of the last test action? something like:

```
 - test:
    definitions:
    - from: inline
      name: v4l2-decoder-conformance-vp9
      path: inline/v4l2-decoder-conformance-vp9.yaml
      repository:
        metadata:
          description: V4L2 decoder conformance test plan
          format: Lava-Test Test Definition 1.0
          name: v4l2-decoder-conformance-vp9
          scope:
          - functional
        run:
          steps:
          - python3 /usr/bin/fluster_parser.py -ts VP9-TEST-VECTORS -d GStreamer-VP9-V4L2SL-Gst1.0
    namespace: chromeos
    timeout:
      minutes: 30 
```

If that works, we can then adjust the template in https://github.com/kernelci/kernelci-core/pull/2617 and pick up from there.

denisyuji commented:
> If that works, we can then adjust the template in [kernelci/kernelci-core#2617](https://github.com/kernelci/kernelci-core/pull/2617) and pick up from there.

Thank you @laura-nao . I simplified the test and removed the downloading/unpacking unnecessary steps. See https://github.com/kernelci/kernelci-core/pull/2617

pawiecz reviewed (approved):
LGTM

I'll lift `staging-skip` status (even though it's Friday) and will also monitor applying this PR to staging because I'd like to have some results ready by Monday

JenySadadia commented:
Hello @pawiecz 
As discussed, we have changed kcidb suite mapping for fluster tests. Please see https://github.com/kernelci/kernelci-pipeline/blob/main/config/jobs-chromeos.yaml#L614.

pawiecz commented:
Thanks @JenySadadia for clarifying that! It looks like this patch still needs to be rebased at least on https://github.com/kernelci/kernelci-pipeline/commit/668466e076a5ed661eb2633d5c1c06d02264505b (if not on the current `main` tip) and adjusted to provide kcidb suite mappings.

denisyuji commented on config/runtime/fluster-chromeos.jinja2:
```diff
@@ -0,0 +1,6 @@
+{%- extends 'chromeos.jinja2' %}
+
+{%- set test_method = 'fluster-chromeos' %}
```
thank you @pawiecz . I reverted the last commit which was unifying templates and I also added the `params: test_method` field

pawiecz replied:
Sure, that works too. Now `test_method` is set twice: (1) in `pipeline.yaml` as `params: test_method=...` and (2) in this template here (which overrides (1) anyway).

Since this template is not used for any other set of Jobs let's leave it as it is now. It can be dealt with (along the template unification) as a follow-up PR once Fluster Jobs are merged.

denisyuji commented:
Thank you for the heads up @JenySadadia . I renamed this test suite as `kcidb_test_suite: fluster.chromeos` ([see](https://github.com/kernelci/kernelci-pipeline/blob/7729325a22d3aaf162fa94b810864c8e9413b7c0/config/jobs-chromeos.yaml#L614)) targeting chromeos to distinguish from `fluster:v4l2` which targets debian. What do you think?

JenySadadia commented:
> Thank you for the heads up @JenySadadia . I renamed this test suite as `kcidb_test_suite: fluster.chromeos` ([see](https://github.com/kernelci/kernelci-pipeline/blob/3a2407c0d33c88612ed9d4597bcaebb1eda36ea5/config/jobs-chromeos.yaml#L393C5-L393C39)) targeting chromeos to distinguish from `fluster:v4l2` which targets debian. What do you think?

Hello @denisyuji 
Thanks for the updates. We have remapped all the KCIDB suite mapping for flusters test to have `fluster.v4l2.<library>_<codec>` convention.
To follow the same convention, we need to use `fluster.chromeos.v4l2.gstreamer_av1` for `fluster-chromeos-av1` and so on for other tests.

pawiecz reviewed (changes requested):
(Changing review status only to highlight that this PR needs KCIDB mapping adjustment - root cause for Friday's staging hiccup has been addressed in the most recent revision)

JenySadadia commented on config/jobs-chromeos.yaml:
```diff
@@ -376,6 +376,22 @@ _anchors:
+        - next
+        - collabora-chromeos-kernel
+        - media
+    kcidb_test_suite: fluster.chromeos
```
Please drop this line.
```suggestion
```

JenySadadia commented on config/jobs-chromeos.yaml:
```diff
@@ -672,6 +688,67 @@ jobs:
 
+  fluster-chromeos-av1:
+    <<: *fluster-chromeos-job
+    params:
+      <<: *fluster-chromeos-params
+      testsuite: 'AV1-TEST-VECTORS'
+      decoders:
+        - 'GStreamer-AV1-V4L2SL-Gst1.0'
```
```suggestion
    params:
      <<: *fluster-chromeos-params
      testsuite: 'AV1-TEST-VECTORS'
      decoders:
        - 'GStreamer-AV1-V4L2SL-Gst1.0'
    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_av1
```

JenySadadia commented on config/jobs-chromeos.yaml:
```diff
@@ -672,6 +688,67 @@ jobs:
+
+  fluster-chromeos-av1-chromium-10bit:
+    <<: *fluster-chromeos-job
+    params:
+      <<: *fluster-chromeos-params
+      testsuite: 'CHROMIUM-10bit-AV1-TEST-VECTORS'
+      decoders:
+        - 'GStreamer-AV1-V4L2SL-Gst1.0'
```
```suggestion
    params:
      <<: *fluster-chromeos-params
      testsuite: 'CHROMIUM-10bit-AV1-TEST-VECTORS'
      decoders:
        - 'GStreamer-AV1-V4L2SL-Gst1.0'
    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_av1_chromium
```

JenySadadia reviewed (changes requested):
Please update the KCIDB mappings as per suggestions. I have added comments for 2 tests. The same would go for other tests as well.

denisyuji commented:
Thanks a lot for the suggestion and changes, @JenySadadia . Can you help me by reviewing it?

JenySadadia commented on config/jobs-chromeos.yaml:
```diff
@@ -672,6 +687,74 @@ jobs:
+      decoders:
+        - 'GStreamer-VP8-V4L2-Gst1.0'
+        - 'GStreamer-VP8-V4L2SL-Gst1.0'
+    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_vp8_chromium
```
```suggestion
    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_vp8
```

JenySadadia commented on config/jobs-chromeos.yaml:
```diff
@@ -672,6 +687,74 @@ jobs:
+      decoders:
+        - 'GStreamer-H.265-V4L2-Gst1.0'
+        - 'GStreamer-H.265-V4L2SL-Gst1.0'
+    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_h265_chromium
```
```suggestion
    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_h265
```

JenySadadia commented on config/jobs-chromeos.yaml:
```diff
@@ -672,6 +687,74 @@ jobs:
+      decoders:
+        - 'GStreamer-H.264-V4L2-Gst1.0'
+        - 'GStreamer-H.264-V4L2SL-Gst1.0'
+    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_h264_frext_chromium
```
```suggestion
    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_h264_frext
```

JenySadadia commented on config/jobs-chromeos.yaml:
```diff
@@ -672,6 +687,74 @@ jobs:
+      decoders:
+        - 'GStreamer-H.264-V4L2-Gst1.0'
+        - 'GStreamer-H.264-V4L2SL-Gst1.0'
+    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_h264_chromium
```
```suggestion
    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_h264
```

JenySadadia commented on config/jobs-chromeos.yaml:
```diff
@@ -672,6 +687,74 @@ jobs:
+      decoders:
+        - 'GStreamer-VP9-V4L2-Gst1.0'
+        - 'GStreamer-VP9-V4L2SL-Gst1.0'
+    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_vp9_chromium
```
```suggestion
    kcidb_test_suite: fluster.chromeos.v4l2.gstreamer_vp9
```

JenySadadia reviewed (changes requested):
Please drop `chromium` suffix from the mapping. It's only required for chromium test suites such as `CHROMIUM-10bit-AV1-TEST-VECTORS`

denisyuji commented:
Thank you @JenySadadia . This is done.

pawiecz reviewed (approved):
LGTM

TODO: unify [`test_method` approaches as described on review](https://github.com/kernelci/kernelci-pipeline/pull/705/commits/bad8fce682ca4691a56a2927dc055e2c291365d3#r1740616620)

(also dropping `staging_skip` label)
